uaclient/api/u/pro/services/disable/v1.py

```diff
@@ -0,0 +1,127 @@
+import logging
+from typing import List, Optional
+
+from uaclient import entitlements, lock, messages, util
+from uaclient.api import AbstractProgress, ProgressWrapper, exceptions
+from uaclient.api.api import APIEndpoint
+from uaclient.api.data_types import AdditionalInfo
+from uaclient.api.u.pro.status.enabled_services.v1 import _enabled_services
+from uaclient.api.u.pro.status.is_attached.v1 import _is_attached
+from uaclient.config import UAConfig
+from uaclient.data_types import (
+    BoolDataValue,
+    DataObject,
+    Field,
+    StringDataValue,
+    data_list,
+)
+
+LOG = logging.getLogger(util.replace_top_level_logger_name(__name__))
+
+
+class DisableOptions(DataObject):
+    fields = [
+        Field("service", StringDataValue),
+        Field("purge", BoolDataValue, False),
+    ]
+
+    def __init__(self, *, service: str, purge: bool = False):
+        self.service = service
+        self.purge = purge
+
+
+class DisableResult(DataObject, AdditionalInfo):
+    fields = [
+        Field("disabled", data_list(StringDataValue)),
+    ]
+
+    def __init__(self, *, disabled: List[str]):
+        self.disabled = disabled
+
+
+def _enabled_services_names(cfg: UAConfig) -> List[str]:
+    return [s.name for s in _enabled_services(cfg).enabled_services]
+
+
+def disable(
+    options: DisableOptions, progress_object: Optional[AbstractProgress] = None
+) -> DisableResult:
+    return _disable(options, UAConfig(), progress_object=progress_object)
+
+
+def _disable(
+    options: DisableOptions,
+    cfg: UAConfig,
+    progress_object: Optional[AbstractProgress] = None,
+) -> DisableResult:
+    progress = ProgressWrapper(progress_object)
+
+    if not util.we_are_currently_root():
+        raise exceptions.NonRootUserError()
+
+    if not _is_attached(cfg).is_attached:
+        raise exceptions.UnattachedError()
+
+    enabled_services_before = _enabled_services_names(cfg)
+    if options.service not in enabled_services_before:
+        # nothing to do
+        return DisableResult(
+            disabled=[],
+        )
+
+    ent_cls = entitlements.entitlement_factory(cfg=cfg, name=options.service)
+    entitlement = ent_cls(
+        cfg,
+        assume_yes=True,
+        called_name=options.service,
+        purge=options.purge,
+    )
```

Comment:
Let's verify if the user provided an invalid service

Reply:
The exception already raised looks pretty good and I added it to the integration test. It looks like this:
```
        {
          "code": "entitlement-not-found",
          "meta": {
            "entitlement_name": "invalid"
          },
          "title": "could not find entitlement named \"invalid\""
        }
```